protocol/messenger_store_node_request_manager.go

```diff
@@ -66,17 +68,45 @@ func (m *StoreNodeRequestManager) FetchCommunity(community communities.Community
 		zap.Any("community", community),
 		zap.Bool("waitForResponse", waitForResponse))
 
-	channel, err := m.subscribeToRequest(storeNodeCommunityRequest, community.CommunityID, community.Shard)
-	if err != nil {
-		return nil, StoreNodeRequestStats{}, fmt.Errorf("failed to create a request for community: %w", err)
+	requestCommunity := func(communityID string, shard *shard.Shard) (*communities.Community, StoreNodeRequestStats, error) {
+		channel, err := m.subscribeToRequest(storeNodeCommunityRequest, communityID, shard)
+		if err != nil {
+			return nil, StoreNodeRequestStats{}, fmt.Errorf("failed to create a request for community: %w", err)
+		}
+
+		if !waitForResponse {
+			return nil, StoreNodeRequestStats{}, nil
+		}
+
+		result := <-channel
+		return result.community, result.stats, result.err
 	}
 
-	if !waitForResponse {
-		return nil, StoreNodeRequestStats{}, nil
+	// if shard was not passed or nil, request shard first
+	communityShard := community.Shard
+	if communityShard == nil {
+		id := transport.CommunityShardInfoTopic(community.CommunityID)
+		shard, err := m.subscribeToRequest(storeNodeShardRequest, id, shard.DefaultShard())
+		if err != nil {
+			return nil, StoreNodeRequestStats{}, fmt.Errorf("failed to create a shard info request: %w", err)
+		}
+
+		if !waitForResponse {
+			go func() {
+				shardResult := <-shard
+				communityShard = shardResult.shard
+
+				_, _, _ = requestCommunity(community.CommunityID, communityShard)
```

Comment:
not an issue: maybe it's possible to early return here?

Reply:
this is executed in a separate thread, so return does not make sence